### PR TITLE
fix hoburg/d8#65

### PR DESCRIPTION
**Pull request by bqpd**
Once I could reproduce, determined #1091 was in fact responsible for hoburg/d8#65 ! Merging once it passes tests to get d8 back on master, but ready for @whoburg review. :)

### CODE REVIEW

File: gpkit/keydict.py [hunk 1]
```diff
@@ -131,6 +140,15 @@ def __getitem__(self, key):
     def __setitem__(self, key, value):
         "Overloads __setitem__ and []= to work with all keys"
         key, idx = self.parse_and_index(key)
+        if hasattr(value, "exp") and not value.exp:
+            value = value.value  # substitute constant monomials
```
Comment (whoburg) [inline gpkit/keydict.py]: this seems surprisingly similar to the first two lines of `clean_value`? Maybe no way around that

File: gpkit/keydict.py [hunk 2]
```diff
@@ -131,6 +140,15 @@ def __getitem__(self, key):
...
+        # pylint: disable=too-many-boolean-expressions
+        if (self.collapse_arrays and hasattr(key, "descr")
+                and "shape" in key.descr and not idx  # if a veckey, not
+                and not isinstance(value, (np.ndarray, Quantity))  # an array,
+                and not is_sweepvar(value)  # not a sweep, and
+                and not isinstance(value[0], np.ndarray)):  # not a solarray,
```
Comment (whoburg) [inline gpkit/keydict.py]: wow

File: gpkit/keydict.py (hunk 1 above)
Comment (bqpd) [inline gpkit/keydict.py]: Yeah, there's this awkward overlap between what's required for a variable value and for an array that might contain variable values. I tried representing that more directly, but this was more readable.

File: gpkit/keydict.py (hunk 2 above)
Comment (bqpd) [inline gpkit/keydict.py]: yeah...something needs to give, here

**Pull request merged**